lib/web_catalog/api_extractor.es6.js

```diff
@@ -5,49 +5,184 @@
+  implements: ['org.chromium.apis.web.ObjectGraphApiMapper'],
 
-  documentation: 'Extract api catalog from a object graph.',
+  documentation: `No-op implementation of ObjectGraphApiMapper.`,
+
+  properties: [
+    ['objectGraph_', {}],
+    ['map_', {}],
+  ],
+
+  methods: [
+    function map(interfaceId, apiNames, sourceId) {},
+    function getObjectGraph() { return this.objectGraph_; },
+    function getMap() { return this.map_; },
+  ],
+});
+
+foam.CLASS({
+  package: 'org.chromium.apis.web',
+  name: 'DebugObjectGraphApiMapper',
+  implements: ['org.chromium.apis.web.ObjectGraphApiMapper'],
+
+  documentation: `Minimal implementation of ObjectGraphApiMapper that returns
+      accurate data from all three interface functions.`,
+
+  properties: [
+    {
+      name: 'objectGraph_',
+      required: true,
+    },
+    ['map_', {}],
+  ],
+
+  methods: [
+    function map(interfaceId, apiNames, sourceId) {
+      let m = this.map_;
+      const interfaceKey = interfaceId.toString();
+      m[interfaceKey] = this.hap_(m, interfaceKey) ? m[interfaceId] : {};
+      for (const apiName of apiNames) {
+        m[interfaceKey][apiName] = sourceId;
+      }
+    },
+    function getObjectGraph() {
+      this.validate();
+      return this.objectGraph_;
+    },
+    function getMap() {
+      this.validate();
+      return this.map_;
+    },
+    function hap_(object, propertyName) {
+      return Object.prototype.hasOwnProperty.call(object, propertyName);
+    },
+  ],
+});
+
+foam.CLASS({
+  package: 'org.chromium.apis.web',
+  name: 'ApiExtractor',
+
+  documentation: `A component that extracts Interface/API pairs from a
+      JavaScript object graph rooted at the global object. This implementation
+      has several configuration properties that affect the behaviour of the main
+      extraction algorithm.`,
 
   requires: [
-    'org.chromium.apis.web.RemoveInterfacesProcessor',
-    'org.chromium.apis.web.RemoveApisProcessor',
-    'org.chromium.apis.web.AddApiProcessor',
+    'org.chromium.apis.web.CopyToInterfaceProcessor',
     'org.chromium.apis.web.CopyToPrototypeProcessor',
+    'org.chromium.apis.web.NullObjectGraphApiMapper',
+    'org.chromium.apis.web.RemoveApisProcessor',
+    'org.chromium.apis.web.RemoveInterfacesProcessor',
+    'org.chromium.apis.web.RemoveObjectGraphRootProcessor',
   ],
 
   properties: [
     {
-      name: 'retainBuiltInMembers',
       class: 'Boolean',
-      documentation: `When retainBuiltInMembers is true,
-        built-in properties of function and object are returned.`,
-      value: false,
+      name: 'functionNamesFromGraphPaths',
+      documentation: `Deduce function names from object graph paths.`,
+      value: true,
     },
     {
-      name: 'retainConstantMembers',
       class: 'Boolean',
-      documentation: `When retainConstantMembers is false
-        constant primitive properties are filtered.`,
-      value: false,
+      name: 'classNamesFromConstructorProperty',
```

Comment:
Design comment: I think for these optional heuristics, a better implementation might be to have a ClassNameHeuristic interface, implemented by ClassNameFromConstructorPropertyHeuristic, ClassNameFromGraphPathsHeuristic, etc. These can then be provided as appropriate, reordered, and potentially be unit tested.

Reply:
We discussed this offline and decided it wasn't worth doing.